world-chain-builder/crates/world/pbh/src/external_nullifier.rs

```diff
@@ -1,121 +1,155 @@
+    pub fn with_date_marker(marker: DateMarker, nonce: u8) -> Self {
+        Self::v1(marker.month as u8, marker.year as u16, nonce)
     }
 
-    pub fn new(prefix: Prefix, date_marker: DateMarker, nonce: u16) -> Self {
+    pub fn v1(month: u8, year: u16, nonce: u8) -> Self {
         Self {
-            prefix,
-            date_marker,
+            version: Prefix::V1,
+            year,
+            month,
             nonce,
         }
     }
 
+    pub fn date_marker(&self) -> DateMarker {
+        DateMarker::new(self.year as i32, self.month as u32)
+    }
+
+    pub fn be_bytes(&self) -> [u8; 32] {
+        let year_bytes = self.year.to_be_bytes();
+
+        let mut bytes = [0; 32];
+
+        bytes[4] = year_bytes[1];
+        bytes[3] = year_bytes[0];
+        bytes[2] = self.month;
+        bytes[1] = self.nonce;
+        bytes[0] = self.version as u8;
+
+        bytes
+    }
+
+    pub fn from_bytes(bytes: [u8; 32]) -> Self {
+        Self::try_from_bytes(bytes).expect("Invalid version")
+    }
+
+    pub fn try_from_bytes(bytes: [u8; 32]) -> Option<Self> {
+        let version = if bytes[0] == Prefix::V1 as u8 {
+            Prefix::V1
+        } else {
+            return None;
+        };
+
+        let mut year_bytes = [0; 2];
+        year_bytes[1] = bytes[4];
+        year_bytes[0] = bytes[3];
+        let year = u16::from_be_bytes(year_bytes);
+
+        let month = bytes[2];
+        let nonce = bytes[1];
+
+        Some(Self {
+            version,
+            year,
+            month,
+            nonce,
+        })
+    }
+
+    pub fn to_word(&self) -> U256 {
+        let bytes = self.be_bytes();
+
+        U256::from_be_bytes(bytes)
+    }
+
+    pub fn from_word(word: U256) -> Self {
+        Self::try_from_word(word).expect("Invalid version")
+    }
+
+    pub fn try_from_word(word: U256) -> Option<Self> {
+        let bytes = word.to_be_bytes();
+
+        Self::try_from_bytes(bytes)
+    }
+
+    // TODO: Rename, this is no longer a hash, but simply the word repr itself
     pub fn hash(&self) -> Field {
-        hash_to_field(self.to_string().as_bytes())
+        self.to_word()
     }
 }
 
 impl std::fmt::Display for ExternalNullifier {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}-{}-{}", self.prefix, self.date_marker, self.nonce)
+        let w = self.to_word();
+        std::fmt::Display::fmt(&w, f)
     }
 }
 
 #[derive(Debug, Clone, PartialEq, Eq, Error)]
 pub enum ExternalNullifierParsingError {
-    #[error("invalid format - expected a string of format `vv-mmyyyy-xxx...` got {actual}")]
-    InvaldFormat { actual: String },
-
-    #[error("error parsing prefix - {0}")]
-    InvalidPrefix(strum::ParseError),
-
-    #[error("error parsing date marker - {0}")]
-    InvalidDateMarker(#[from] DateMarkerParsingError),
-
-    #[error("error parsing nonce - {0}")]
-    InvalidNonce(std::num::ParseIntError),
```

Comment:
LGTM! Only suggestion would be to leave in the granular errors! Might be nice to have.

Reply:
Good point I'll re-add month validation

Reply:
Other errors don't really make sense at "parse time"